regression: PCH with msvc causing "error LNK2011: precompiled object not linked in; image may not run"

Hi,

As mentioned on IRC: Starting with @dcbaker 's c94c492089e7fecb56b7cc11ea76712a770f1e34 I see msvc builds of postgres fail when pch is enabled:
```
[56/56] Linking target src/bin/pg_config/pg_config.exe
FAILED: src/bin/pg_config/pg_config.exe src/bin/pg_config/pg_config.pdb
"link"  /MACHINE:x64 /OUT:src/bin/pg_config/pg_config.exe src/bin/pg_config/pg_config.exe.p/pg_config.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:src\bin\pg_config\pg_config.pdb" "src/common\libpgcommon.a" "src/fe_utils\libpgfeutils.a" "src/port\libpgport.a" "ws2_32.lib" "ws2_32.lib" "ws2_32.lib" "ws2_32.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
libpgcommon.a(config_info.c.obj) : error LNK2011: precompiled object not linked in; image may not run
  Hint on symbols that are defined and could potentially match:
    __@@_PchSym_@00@UwveUklhgtivhRnvhlmUyfrowRmrmqzUhixUkligUoryktkligOzOkUxPkxsOlyq@pgport_main
    __@@_PchSym_@00@UwveUklhgtivhRnvhlmUyfrowRmrmqzUhixUxlnnlmUoryktxlnnlmOzOkUxPkxsOlyq@pgcommon
```

The @pgport_main etc is the result of me going through the static libraries and naming them with `/Yl<somename>` as part of trying to debug this.

I spent a decent bit of time trying to build a reproducer, but I'm getting too tired.  I suspect it's related to a static library that links to another static library. But it doesn't trigger in a simplistic reproducer unfortunately.

Regards,

Andres

Just when I was ready to give up I reproduced it. For some reason the example needs link_whole, while postgres's case doesn't need link_whole. Assume somehow postgres's link_with gets transformed to a link_whole.

https://github.com/anarazel/meson/commit/90bfa3fdf4b6cedef9683d56c9ec3d4bb8b8a091 hacks the existing test up to show the problem. There's lots of gunk in there that needs to be cleaned up, but that's for tomorrow.

```
PS C:\dev\meson\test cases\common\13 pch\build> ninja -v c/prog.exe        
[1/12] "cl" "-Ic\libstlib_squared.a.p" "-Ic" "-I..\c" "-I..\c\pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/Ycstatic_squared_pch.h" "/Fpc\libstlib_squared.a.p\static_squared_pch.pch" "/Foc\libstlib_squared.a.p\static_squared_pch.obj" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\libstlib_squared.pdb" "/c" c/libstlib_squared.a.p/meson_pch-c.c
[2/12] "cl" "-Ic\shared.dll.p" "-Ic" "-I..\c" "-I..\c\pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/Ycshared_pch.h" "/Fpc\shared.dll.p\shared_pch.pch" "/Foc\shared.dll.p\shared_pch.obj" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "-DBUILDING_DLL" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\shared.pdb" "/c" c/shared.dll.p/meson_pch-c.c
[3/12] "cl" "-Ic\libstatic.a.p" "-Ic" "-I..\c" "-I..\c\pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/Ycstatic_pch.h" "/Fpc\libstatic.a.p\static_pch.pch" "/Foc\libstatic.a.p\static_pch.obj" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" /FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\libstatic.pdb" "/c" c/libstatic.a.p/meson_pch-c.c
[4/12] "cl" "-Ic\prog.exe.p" "-Ic" "-I..\c" "-I..\c\pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/Ycprog.h" "/Fpc\prog.exe.pprog.pch" "/Foc\prog.exe.p\prog.obj" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\devmeson\test cases\common\13 pch\build\c\prog.pdb" "/c" c/prog.exe.p/meson_pch-c.c
[5/12] "cl" "-Ic\shared.dll.p" "-Ic" "-I..\c" "/MDd" "/FIshared_pch.h" "/Yushared_pch.h" "/Fpc\shared.dll.p\shared_pch.pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "-DBUILDING_DLL" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\buildcshared.pdb" /Foc/shared.dll.p/shared.c.obj "/c" ../c/shared.c       
[6/12] "cl" "-Ic\libstlib_squared.a.p" "-Ic" "-I..\c" "/MDd" "/FIstatic_squared_pch.h" "/Yustatic_squared_pch.h" "/Fpc\libstlib_squared.a.p\static_squared_pch.pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\libstlib_squared.pdb" /Foc/libstlib_squared.a.p/static_squared.c.obj "/c" ../c/static_squared.c
[7/12] "cl" "-Ic\libstatic.a.p" "-Ic" "-I..\c" "/MDd" "/FIstatic_pch.h" "/Yustatic_pch.h" "/Fpc\libstatic.a.p\static_pch.pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\libstatic.pdb" /Foc/libstatic.a.p/static.c.obj "/c" ../c/static.c
[8/12] "cl" "-Ic\prog.exe.p" "-Ic" "-I..\c" "/MDd" "/FIprog.h" "/Yuprog.h" "/Fpc\prog.exe.p\prog.pch" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/FS" "/FdC:\dev\meson\test cases\common\13 pch\build\c\prog.pdb" /Foc/prog.exe.p/prog.c.obj "/c" ../c/prog.c
[9/12] "lib" "/NOLOGO" "c\libstatic.a.p\static_pch.obj" /MACHINE:x64 /OUT:c/libstatic.a c/libstlib_squared.a.p/static_squared.c.obj c/libstatic.a.p/static.c.obj
[10/12] "link"  /MACHINE:x64 /OUT:c/shared.dll c/shared.dll.p/shared.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:c\shared.pdb" "/DLL" "/IMPLIB:c\shared.lib" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib" "c\shared.dll.p\shared_pch.obj"
   Creating library c\shared.lib and object c\shared.exp
[11/12] "C:\python\python.exe" "C:\dev\meson\meson.py" "--internal" "symbolextractor" "C:\dev\meson\test cases\common\13 pch\build" c/shared.dll "c\shared.lib" c/shared.dll.p/shared.dll.symbols
[12/12] "link"  /MACHINE:x64 /OUT:c/prog.exe c/prog.exe.p/prog.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:c\prog.pdb" "c\libstatic.a" "c\shared.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib" "c\prog.exe.p\prog.obj"
FAILED: c/prog.exe c/prog.pdb
"link"  /MACHINE:x64 /OUT:c/prog.exe c/prog.exe.p/prog.c.obj "/nologo" "/release" "/nologo" "/DEBUG" "/PDB:c\prog.pdb" "c\libstatic.a" "c\shared.lib" "/SUBSYST
EM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib" "c\prog.ex
e.p\prog.obj"
libstatic.a(static_squared.c.obj) : error LNK2011: precompiled object not linked in; image may not run
  Hint on symbols that are defined and could potentially match:
    __@@_PchSym_@00@UwveUnvhlmUgvhgLxzhvhUxlnnlmUBDLkxsUyfrowUxUkiltOvcvOkUkiltOlyq@4B2008FD98C1DD4
    __@@_PchSym_@00@UwveUnvhlmUgvhgLxzhvhUxlnnlmUBDLkxsUyfrowUxUoryhgzgrxOzOkUhgzgrxPkxsOlyq@4B2008FD98C1DD4
c\prog.exe : fatal error LNK1120: 1 unresolved externals
ninja: build stopped: subcommand failed.
```

The problem is
```
[9/12] "lib" "/NOLOGO" "c\libstatic.a.p\static_pch.obj" /MACHINE:x64 /OUT:c/libstatic.a c/libstlib_squared.a.p/static_squared.c.obj c/libstatic.a.p/static.c.obj
```
which is omitting c\libstlib_squared.a.p\static_squared_pch.obj, generated in [1/12].

Looks like CI in my own repo won't run windows tests? I'll open a draft PR to show the problem...

Regards,

Andres

--- a/mesonbuild/backend/ninjabackend.py
+++ b/mesonbuild/backend/ninjabackend.py
@@ -37,7 +37,7 @@ from .. import mlog
 from .. import compilers
 from ..arglist import CompilerArgs
 from ..compilers import Compiler
-from ..linkers import ArLinker, AppleArLinker, RSPFileSyntax
+from ..linkers import ArLinker, RSPFileSyntax
 from ..mesonlib import (
     File, LibType, MachineChoice, MesonException, OrderedSet, PerMachine,
     ProgressBar, quote_arg
@@ -2139,7 +2139,7 @@ class NinjaBackend(backends.Backend):
             if static_linker is None:
                 continue
             rule = 'STATIC_LINKER{}'.format(self.get_rule_suffix(for_machine))
-            cmdlist: T.List[T.Union[str, NinjaCommandArg]] = []
+            cmdlist = []
             args = ['$in']
             # FIXME: Must normalize file names with pathlib.Path before writing
             #        them out to fix this properly on Windows. See:
@@ -2153,17 +2153,6 @@ class NinjaBackend(backends.Backend):
             cmdlist += static_linker.get_exelist()
             cmdlist += ['$LINK_ARGS']
             cmdlist += NinjaCommandArg.list(static_linker.get_output_args('$out'), Quoting.none)
-            # The default ar on MacOS (at least through version 12), does not
-            # add extern'd variables to the symbol table by default, and
-            # requires that apple's ranlib be called with a special flag
-            # instead after linking
-            if isinstance(static_linker, AppleArLinker):
-                # This is a bit of a hack, but we assume that that we won't need
-                # an rspfile on MacOS, otherwise the arguments are passed to
-                # ranlib, not to ar
-                cmdlist.extend(args)
-                args = []
-                cmdlist.extend(['&&', 'ranlib', '-c', '$out'])
             description = 'Linking static target $out'
             if num_pools > 0:
                 pool = 'pool = link_pool'

--- a/mesonbuild/build.py
+++ b/mesonbuild/build.py
@@ -1403,13 +1403,11 @@ You probably should put it in link_with instead.''')
                     raise InvalidArguments(msg + ' This is not possible in a cross build.')
                 else:
                     mlog.warning(msg + ' This will fail in cross build.')
-            # When we're a static library and we link_whole: to another static
-            # library, we need to add that target's objects to ourselves.
-            # Otherwise add it to the link_whole_targets, but not both.
             if isinstance(self, StaticLibrary):
+                # When we're a static library and we link_whole: to another static
+                # library, we need to add that target's objects to ourselves.
                 self.objects += t.extract_all_objects_recurse()
-            else:
-                self.link_whole_targets.append(t)
+            self.link_whole_targets.append(t)
 
     def extract_all_objects_recurse(self) -> T.List[T.Union[str, 'ExtractedObjects']]:
         objs = [self.extract_all_objects()]

--- a/test cases/osx/9 global variable ar/libfile.c
+++ b/test cases/osx/9 global variable ar/libfile.c
@@ -1,9 +1,0 @@
-// Source: https://lists.gnu.org/archive/html/libtool/2002-07/msg00025.html
-
-#include <stdio.h>
-
-extern int l2;
-void l1(void)
-{
-  printf("l1 %d\n", l2);
-}

--- a/test cases/osx/9 global variable ar/libfile2.c
+++ b/test cases/osx/9 global variable ar/libfile2.c
@@ -1,7 +1,0 @@
-// Source: https://lists.gnu.org/archive/html/libtool/2002-07/msg00025.html
-
-int l2;
-void l2_func(void)
-{
-  l2 = 77;
-}

--- a/test cases/osx/9 global variable ar/meson.build
+++ b/test cases/osx/9 global variable ar/meson.build
@@ -1,6 +1,0 @@
-# Source: https://lists.gnu.org/archive/html/libtool/2002-07/msg00025.html
-
-project('global variable test', 'c')
-
-lib = static_library('mylib', 'libfile.c', 'libfile2.c')
-test('global variable', executable('prog', 'prog.c', link_with: lib))

--- a/test cases/osx/9 global variable ar/nativefile.ini
+++ b/test cases/osx/9 global variable ar/nativefile.ini
@@ -1,2 +1,0 @@
-[binaries]
-ar = 'ar'

--- a/test cases/osx/9 global variable ar/prog.c
+++ b/test cases/osx/9 global variable ar/prog.c
@@ -1,7 +1,0 @@
-// Source: https://lists.gnu.org/archive/html/libtool/2002-07/msg00025.html
-
-extern void l1(void);
-int main(void)
-{
-  l1();
-}

